pkg/compact/downsample/downsample.go

```diff
@@ -0,0 +1,585 @@
+	// chunks align with boundaries of input chunks.
+	// Otherwise the signal sample indiciating the last true sample value is not picked correctly.
+	n := targetChunkSize(len(data))
+	k := len(chunked)/(len(data)/n) + 1 // batch size
+
+	for len(chunked) > 0 {
+		j := k
+		if j > len(chunked) {
+			j = len(chunked)
+		}
+		length := 0
+		for _, p := range chunked[:j] {
+			length += len(p)
+		}
+		part := data[:length]
+		data = data[length:]
+		chunked = chunked[j:]
+
+		// Downsample the counter series spanning the chunk batch into a single chunk.
+		downsampleCounter(ser, part, window, false)
+	}
+}
+
+func downsampleCount(ser *series, data []sample, window int64) {
+	nextT := int64(-1)
+	var count int
+
+	for _, s := range data {
+		if s.t > nextT {
+			if nextT != -1 {
+				ser.add(nextT, float64(count))
+			}
+			count = 0
+			nextT = s.t - (s.t % window) + window - 1
+		}
+		count++
+	}
+	ser.add(nextT, float64(count))
+}
+
+func downsampleSum(ser *series, data []sample, window int64) {
+	nextT := int64(-1)
+	var sum float64
+
+	for _, s := range data {
+		if s.t > nextT {
+			if nextT != -1 {
+				ser.add(nextT, sum)
+			}
+			sum = 0
+			nextT = s.t - (s.t % window) + window - 1
+		}
+		sum += s.v
+	}
+	ser.add(nextT, sum)
+}
+
+func downsampleMin(ser *series, data []sample, window int64) {
+	nextT := int64(-1)
+	min := math.MaxFloat64
+
+	for _, s := range data {
+		if s.t > nextT {
+			if nextT != -1 {
+				ser.add(nextT, min)
+			}
+			min = math.MaxFloat64
+			nextT = s.t - (s.t % window) + window - 1
+		}
+		if s.v < min {
+			min = s.v
+		}
+	}
+	ser.add(nextT, min)
+}
+
+func downsampleMax(ser *series, data []sample, window int64) {
+	nextT := int64(-1)
+	max := -math.MaxFloat64
+
+	for _, s := range data {
+		if s.t > nextT {
+			if nextT != -1 {
+				ser.add(nextT, max)
+			}
+			max = -math.MaxFloat64
+			nextT = s.t - (s.t % window) + window - 1
+		}
+		if s.v > max {
+			max = s.v
+		}
+	}
+	ser.add(nextT, max)
+}
+
+// countChunkSeriesIterator iterates over an ordered sequence of chunks and treats decreasing
+// values as counter reset.
+// Additionally, it can deal with downsampled counter chunks, which set the last value of a chunk
+// to the original last value. The last value can be detected by checking whether the timestamp
+// did not increase w.r.t to the previous sample
+type countChunkSeriesIterator struct {
+	chks   []chunkenc.Iterator
+	i      int     // current chunk
+	total  int     // total number of processed samples
+	lastT  int64   // timestamp of the last sample
+	lastV  float64 // value of the last sample
+	totalV float64 // total counter state since beginning of series
+}
+
+func (it *countChunkSeriesIterator) LastSample() (int64, float64) {
+	return it.lastT, it.lastV
+}
+
+func (it *countChunkSeriesIterator) Next() bool {
+	if it.i >= len(it.chks) {
+		return false
+	}
+	// Chunk ends without special sample. It was not a downsampled counter series.
```

Comment:
Is this comment relevant? It feels out of context

Reply:
Will clarify.